crates/plugins/mediator-coordination/src/forward/routing.rs

```diff
@@ -0,0 +1,132 @@
+        for did in dids {
+            let (re_packed_msg, _) = unpack_msg
+                .pack_encrypted(
+                    &did,
+                    Some(mediator_did),
+                    None,
+                    &state.did_resolver,
+                    &state.secrets_resolver,
+                    &PackEncryptedOptions::default(),
+                )
+                .await
+                .expect("could not pack message: {0}");
+
+            let messages = Messages {
+                id: None,
+                message: unpack_msg.clone(),
+                recipient_did: did,
+            };
+            message_repository
+                .store(messages)
+                .await
+                .map_err(|_| MediationError::PersisenceError)
+                .unwrap();
+            forward.push(re_packed_msg)
+        }
+    }
+
+    Ok(forward)
+}
+
+#[cfg(test)]
+mod test {
+    use std::sync::Arc;
+
+    use crate::{
+        repository::stateful::coord::tests::{
+            MockConnectionRepository, MockMessagesRepository, MockSecretsRepository,
+        },
+        util::{self, MockFileSystem},
+        web::AppStateRepository,
+    };
+
+    use super::*;
+
+    use didcomm::Message;
+    pub fn setup() -> Arc<AppState> {
+        let public_domain = String::from("http://alice-mediator.com");
+
+        let mut mock_fs = MockFileSystem;
+        let diddoc = util::read_diddoc(&mock_fs, "").unwrap();
+        let keystore = util::read_keystore(&mut mock_fs, "").unwrap();
+
+        let repository = AppStateRepository {
+            connection_repository: Arc::new(MockConnectionRepository::from(vec![])),
+            secret_repository: Arc::new(MockSecretsRepository::from(vec![])),
+            message_repository: Arc::new(MockMessagesRepository::from(vec![])),
+        };
+
+        let state = Arc::new(AppState::from(
+            public_domain,
+            diddoc,
+            keystore,
+            Some(repository),
+        ));
+
+        state
+    }
+    #[tokio::test]
+    /// simulate sender forwarding process
+    async fn test_mediator_forward_process() {
+        let msg: Message = Message::build(
+            "id".to_owned(),
+            "type_".to_owned(),
+            serde_json::json!("example-body"),
+        )
+        .to("did:key:z6MkhaXgBZDvotDkL5257faiztiGiC2QtKLGpbnnEGta2doK".to_owned())
+        .from("did:web:alice-mediator.com:alice_mediator_pub".to_owned())
+        .finalize();
+
+        let state = setup();
+        let (msg, _metadata) = msg
+            .pack_encrypted(
+                "did:key:z6MkhaXgBZDvotDkL5257faiztiGiC2QtKLGpbnnEGta2doK",
+                Some("did:web:alice-mediator.com:alice_mediator_pub"),
+                None,
+                &state.did_resolver,
+                &state.secrets_resolver,
+                &PackEncryptedOptions::default(),
+            )
+            .await
+            .unwrap();
```

Comment:
Same, the original sender is not the mediator. Also, you carefully need to prepare a **forward** message. 

I recommend to clone this crate we're using: https://github.com/sicpa-dlab/didcomm-rust. Then run `routing.rs` in the `examples` folder. You'll have an idea of a complete forward flow.

You'll also find my changes to the `single_mediator()` scenario useful.

```rust
#[allow(unused_imports, dead_code)]
#[path = "../src/test_vectors/mod.rs"]
mod test_vectors;

// TODO: look for better solution
// Allows test vectors usage inside and outside crate
pub(crate) use didcomm;

use didcomm::{
    algorithms::AnonCryptAlg,
    did::resolvers::ExampleDIDResolver,
    protocols::routing::{self, try_parse_forward},
    secrets::resolvers::ExampleSecretsResolver,
    Message, PackEncryptedOptions, UnpackOptions,
};
use serde_json::json;
use test_vectors::{
    ALICE_DID, ALICE_DID_DOC, ALICE_SECRETS, BOB_DID, BOB_DID_DOC, BOB_SECRETS, MEDIATOR1_DID_DOC,
    MEDIATOR1_SECRETS,
};

#[tokio::main(flavor = "current_thread")]
async fn main() {
    single_mediator().await;
}

async fn single_mediator() {
    // --- Building message from ALICE to BOB ---
    let msg = Message::build(
        "example-1".to_owned(),
        "example/v1".to_owned(),
        json!("example-body"),
    )
    .to(BOB_DID.to_owned())
    .from(ALICE_DID.to_owned())
    .finalize();

    // --- Packing encrypted and authenticated message ---
    let did_resolver = ExampleDIDResolver::new(vec![
        ALICE_DID_DOC.clone(),
        BOB_DID_DOC.clone(),
        MEDIATOR1_DID_DOC.clone(),
    ]);

    let secrets_resolver = ExampleSecretsResolver::new(ALICE_SECRETS.clone());

    let (msg, metadata) = msg
        .pack_encrypted(
            BOB_DID,
            Some(ALICE_DID),
            None,
            &did_resolver,
            &secrets_resolver,
            &PackEncryptedOptions {
                forward: false,
                ..Default::default()
            },
        )
        .await
        .expect("Unable pack_encrypted");

    println!("Encryption metadata is\n{:?}\n", metadata);

    // -- Wrap in forward

    let msg = routing::wrap_in_forward(
        &msg,
        None,
        BOB_DID,
        &vec![MEDIATOR1_DID_DOC.clone().id],
        &AnonCryptAlg::default(),
        &did_resolver,
    )
    .await
    .expect("Unable wrapping");

    // --- Sending message by Alice ---
    println!("Alice is sending message \n{}\n", msg);

    // --- Unpacking message by Mediator1 ---
    let did_resolver = ExampleDIDResolver::new(vec![
        ALICE_DID_DOC.clone(),
        BOB_DID_DOC.clone(),
        MEDIATOR1_DID_DOC.clone(),
    ]);

    let secrets_resolver = ExampleSecretsResolver::new(MEDIATOR1_SECRETS.clone());

    let (msg, metadata) = Message::unpack(
        &msg,
        &did_resolver,
        &secrets_resolver,
        &UnpackOptions::default(),
    )
    .await
    .expect("Unable unpack");

    println!("Mediator1 received message is \n{:?}\n", msg);

    println!(
        "Mediator1 received message unpack metadata is \n{:?}\n",
        metadata
    );

    // --- Forwarding message by Mediator1 ---
    let msg = serde_json::to_string(&try_parse_forward(&msg).unwrap().forwarded_msg).unwrap();

    println!("Mediator1 is forwarding message \n{}\n", msg);

    // --- Unpacking message by Bob ---
    let did_resolver = ExampleDIDResolver::new(vec![
        ALICE_DID_DOC.clone(),
        BOB_DID_DOC.clone(),
        MEDIATOR1_DID_DOC.clone(),
    ]);

    let secrets_resolver = ExampleSecretsResolver::new(BOB_SECRETS.clone());

    let (msg, metadata) = Message::unpack(
        &msg,
        &did_resolver,
        &secrets_resolver,
        &UnpackOptions::default(),
    )
    .await
    .expect("Unable unpack");

    println!("Bob received message is \n{:?}\n", msg);
    println!("Bob received message unpack metadata is \n{:?}\n", metadata);
}
```

Reply:
Yeah this really helped thanks :pray: